modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/UpdatableTableImpl.java

```diff
@@ -249,59 +252,53 @@ public <RowT> CompletableFuture<?> insertAll(
-
-        int batchNum = 0;
-
-        IgniteTypeFactory typeFactory = ectx.getTypeFactory();
-        RowSchema rowSchema = rowSchemaFromRelTypes(RelOptUtil.getFieldTypeList(desc.insertRowType(typeFactory)));
-        RowHandler.RowFactory<RowT> rowFactory = handler.factory(rowSchema);
+        for (Int2ObjectMap.Entry<RowBatch> partitionRowBatch : rowBatchByPartitionId.int2ObjectEntrySet()) {
+            int partitionId = partitionRowBatch.getIntKey();
+            RowBatch rowBatch = partitionRowBatch.getValue();
 
-        for (Int2ObjectMap.Entry<List<BinaryRow>> partToRows : rowsByPartition.int2ObjectEntrySet()) {
-            TablePartitionId partGroupId = new TablePartitionId(tableId, partToRows.getIntKey());
+            TablePartitionId partGroupId = new TablePartitionId(tableId, partitionId);
             ColocationGroup group = ectx.group(MODIFY_NODE_SOURCE_ID);
 
             assert group != null;
 
-            NodeWithTerm nodeWithTerm = group.assignments().get(partToRows.getIntKey());
+            NodeWithTerm nodeWithTerm = group.assignments().get(partitionId);
 
             ReplicaRequest request = MESSAGES_FACTORY.readWriteMultiRowReplicaRequest()
                     .groupId(partGroupId)
                     .commitPartitionId(serializeTablePartitionId(commitPartitionId))
-                    .binaryRowMessages(serializeBinaryRows(partToRows.getValue()))
+                    .binaryRowMessages(serializeBinaryRows(rowBatch.requestedRows))
                     .transactionId(txAttributes.id())
                     .term(nodeWithTerm.term())
                     .requestType(RequestType.RW_INSERT_ALL)
                     .timestampLong(clock.nowLong())
                     .build();
 
-            futures[batchNum++] = replicaService.invoke(nodeWithTerm.name(), request)
-                    .thenApply(result -> {
-                        Collection<BinaryRow> binaryRows = (Collection<BinaryRow>) result;
+            rowBatch.resultFuture = replicaService.invoke(nodeWithTerm.name(), request);
+        }
+
+        return handleInsertResults(ectx, rowBatchByPartitionId.values());
+    }
 
-                        if (binaryRows.isEmpty()) {
-                            return List.of();
-                        }
+    /**
+     * Creates batches of rows for processing, grouped by partition ID.
+     *
+     * @param rowHandler Row handler.
+     * @param rows Rows.
+     */
+    private <T> Int2ObjectMap<RowBatch> toRowBatchByPartitionId(RowHandler<T> rowHandler, List<T> rows) {
+        Int2ObjectMap<RowBatch> rowBatchByPartitionId = new Int2ObjectOpenHashMap<>();
 
-                        List<RowT> conflictRows = new ArrayList<>(binaryRows.size());
+        int i = 0;
 
-                        for (BinaryRow row : binaryRows) {
-                            conflictRows.add(rowConverter.toRow(ectx, row, rowFactory));
-                        }
+        for (T row : rows) {
```

Comment:
U can use for(int i=0; ...

Reply:
I know.

Reply:
In your case, it will be more effective, one less allocation plus various kinds of optimization.

Reply:
Honestly, I do not understand what you mean. If you think about minor improvements, I do not see a reason for them just in this PR.

Reply:
Ok